### PR TITLE
Make orbit_case::Append more generic

### PR DESCRIPTION
**Pull request by beckerhe**
`Append` used to only accept `const std::vector<T>&` as the source of elements that should be appended. This change makes `Append` also work with other sources - notably with `absl::Span<T const>`.

This is in preparation to replacing all usages of `const std::vector<T>&` as function parameter types by `absl::Span<T const>`.

### CODE REVIEW

File: src/OrbitBase/include/OrbitBase/Append.h [hunk 1]
```diff
@@ -2,16 +2,32 @@
...
-void Append(std::vector<T>& dest, const std::vector<T>& source) {
-  dest.insert(std::end(dest), std::begin(source), std::end(source));
+// Append adds the elements given by the range in the second argument to the back of the vector
+// given in the first argument. The range can be anything that defines `Range::value_type` and has
+// cbegin and cend iterators.
+template <class T, class Range,
+          class = typename std::enable_if_t<std::is_convertible_v<typename Range::value_type, T>>>
+void Append(std::vector<T>& dest, const Range& source) {
+  using std::cbegin;
+  using std::cend;
```
Comment (dpallotti) [inline src/OrbitBase/include/OrbitBase/Append.h:21]: I guess these `using`s are the fallback in the lookup of `cbegin` and `cend` in `insert`? Or am I off track?

File: src/OrbitBase/include/OrbitBase/Append.h [hunk 2]
```diff
@@ -2,16 +2,32 @@
...
 
 namespace orbit_base {
 
-template <class T>
-void Append(std::vector<T>& dest, const std::vector<T>& source) {
-  dest.insert(std::end(dest), std::begin(source), std::end(source));
+// Append adds the elements given by the range in the second argument to the back of the vector
+// given in the first argument. The range can be anything that defines `Range::value_type` and has
+// cbegin and cend iterators.
+template <class T, class Range,
```
Comment (dpallotti) [inline src/OrbitBase/include/OrbitBase/Append.h:17]: I generally prefer `typename` to `class` as `class` is a bit misleading.

File: src/OrbitBase/include/OrbitBase/Append.h [hunk 3]
```diff
@@ -2,16 +2,32 @@
...
+          class = typename std::enable_if_t<std::is_convertible_v<typename Range::value_type, T>>>
+void Append(std::vector<T>& dest, const Range& source) {
+  using std::cbegin;
+  using std::cend;
+  dest.insert(std::end(dest), cbegin(source), cend(source));
+}
+
+// This overload makes initializer lists as the second argument work.
+template <class T, class U, class = typename std::enable_if_t<std::is_convertible_v<U, T>>>
+void Append(std::vector<T>& dest, std::initializer_list<U> source) {
```
Comment (dpallotti) [inline src/OrbitBase/include/OrbitBase/Append.h:27]: You could rearrange the tests in `AppendTest.cpp` a bit distinguishing between `SourceIsVectorVariable`, `SourceIsTemporaryVector`, and `SourceIsTemporaryInitializerList`.

**Review by dpallotti — APPROVED** [review]

File: src/OrbitBase/include/OrbitBase/Append.h (hunk 1 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Append.h:21]: Yes. Typical meta programming shenanigans. Assuming Range is of type `my_ns::Range`, then the compiler will first look for `cbegin` in `my_ns` (ADL) and if not found fall back to `std::cbegin`. `std::cbegin` usually tries to call the member function `cbegin`, except in some cases that don't have member functions like C-style arrays.

File: src/OrbitBase/include/OrbitBase/Append.h (hunk 2 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Append.h]: Done.

File: src/OrbitBase/include/OrbitBase/Append.h (hunk 3 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Append.h:28]: Done.

File: src/OrbitBase/include/OrbitBase/Append.h (hunk 1 above)
Comment (dpallotti) [reply src/OrbitBase/include/OrbitBase/Append.h:22]: Thanks! I'm happy I got it somewhat right :D

**Pull request merged**